utils/dwarf.c

```diff
@@ -263,9 +266,6 @@ static void release_dwarf_info(struct debug_info *dinfo)
 		free(df->name);
 		free(df);
 	}
-
-	free(dinfo->locs);
-	dinfo->locs = NULL;
```

Comment:
The `dinfo->files` also needs to be freed.  I suggest moving both (files + locs) into `release_debug_info()` since they can be used even without DWARF (once recorded).

Reply:
@namhyung As you said, `dinfo->files` also need to be freed. But, cleaning-up `dinfo->files` when `dinfo->dw` is `NULL` incurs CRASH (segmentation fault).

Reply:
The cause of the crash seems that sometimes `dinfo->files` isn't initialized using `INIT_LIST_HEAD`.

Reply:
Hmm.. strange.  It's initialized both from `setup_debug_info()` (for record) and `load_debug_file()` (for replay).

Reply:
But coredump shows that this problem happens from record.

Reply:
And I think we need to initialize `locs` from `setup_debug_info()` as well.

Reply:
Do you have a data causes a crash during replay?  If so, could you please share it?

Reply:
@namhyung Here it is.
The source code:
```cpp
#include <vector>
int main() {
    std::vector<int> vec;
    vec.push_back(12);
}
```
Gcc Command : `gcc -pg vector.c -o vector.out
Uftrace Command : `uftrace vector.out`

Reply:
@namhyung And, sorry to my past saying:
> And if I remove `close_data_file(opts, &handle);` in line 1108 of `cmds/replay.c`, the problem is resolved.

This is wrong. The causes of the problem are two things. One is the thing above. And the another is caused by `finish_debug_info(&symtabs);` in line 1411 of `libmcount/mcount.c`.

Reply:
The another core dump shows below:
```
#0  0x000055d1c99cc995 in release_debug_info (dinfo=0x55d1cae30568)
    at /home/taeguk/uftrace/utils/dwarf.c:1046
#1  0x000055d1c99cce25 in finish_debug_info (symtabs=0x55d1cae49c88)
    at /home/taeguk/uftrace/utils/dwarf.c:1153
#2  0x000055d1c99bf02c in delete_session (sess=0x55d1cae49c50)
    at /home/taeguk/uftrace/utils/session.c:342
#3  0x000055d1c99bf0de in delete_sessions (sessions=0x7ffca1afedd8)
    at /home/taeguk/uftrace/utils/session.c:369
#4  0x000055d1c99c9d64 in close_data_file (opts=0x7ffca1afefd0, handle=0x7ffca1afec80)
    at /home/taeguk/uftrace/utils/data-file.c:539
#5  0x000055d1c99be2a0 in command_replay (argc=1, argv=0x7ffca1aff200, opts=0x7ffca1afefd0)
    at /home/taeguk/uftrace/cmds/replay.c:1108
#6  0x000055d1c99baaa4 in command_live (argc=1, argv=0x7ffca1aff200, opts=0x7ffca1afefd0)
    at /home/taeguk/uftrace/cmds/live.c:147
#7  0x000055d1c999c7ed in main (argc=1, argv=0x7ffca1aff200) at /home/taeguk/uftrace/uftrace.c:1050
```

Reply:
@namhyung Finally, I fixed the crash problem!
https://github.com/namhyung/uftrace/pull/467/commits/cfb3dffe191f404bb43a6cd80d16fed604563bef fixes the crash problem :)

Reply:
Great!  Could you please reorder the commit so that we can have the fix before the change in `release_dwarf_info()`?

Reply:
But, github shows the strange order of commits. Maybe it's the bug of github.